### ISSUE TITLE
for loop can't use an existing variable

### ISSUE THREAD

**Issue by ghost** [fish-shell/fish-shell#1935]
``` fish
set index 1; for index in (seq 10)
  echo $index
end; echo $index
```

```
1
2
3
4
5
6
7
8
9
10
1
```

I thought `$index` would exit the `for` loop with 10 since I declared `$index` outside the loop. Why?

**Comment by zanchey** [comment]
There's a problem with the scoping here; my reading of the documentation says that it should behave as you expect.

**Comment by bucaran** [comment]
In the meantime this is a nice workaround.
```fish
  for _index in (seq $index $count)
    set index $_index
    ...
  end
```

**Comment by ridiculousfish** [comment]
for loops have historically used local scope like `set -l` for the iteration variable. The trouble is that they do this after the new scope has been introduced, that is, the for block is pushed on the stack, so the iteration variable is local to that block. It's been that way since fish 1.x, but I think I agree with bucaran and zanchey that the variable should live in the outer scope.

**Comment by xfix** [comment]
`for` loop declares its iteration variable in its own scope. It works similarly to Perl or Python 3 (but not 2, the behavior has changed in Python 3) in that respect. I wouldn't say it's a bug, because to be honest, behaviour of reusing scope variable in `for` loop can be a potential program bug (if you let's say, accidentally reuse identifier in a loop).

```perl
use 5.012;
my $i = 5;
for $i (1..3) {
    say $i;
}
say $i;
```

```python
>>> i = 5
>>> [i for i in range(3)]
[0, 1, 2]
>>> i
5
>>>
```

As for workaround if you need last value in the loop, it's quite easy. Just assign it to unscoped variable.

```fish
set -l last
for index in (seq 10)
    set last $index
    echo $index
end
echo $last
```

**Comment by bucaran** [comment]
@xfix Bug or not, it would be nice if I could use the same iterator variable after the loop has ended, granted the variable was declared outside the scope of the loop. 

I would like to propose the following new syntax:

```fish
for set -l index 1 in (seq 10)
  echo $index
end
```

That would be the default behavior, which is the most common use case, but removing the `-l` would make `index` global to the current function or scope and allow me to do away with any workarounds.

This would be similar to:

```fish
if set -l result (get_data)
  # ...
end
```

**Comment by GReagle** [comment]
I know that fish is not a Bourne shell and doesn't have to be compatible with anything else.  However, I think it is useful to compare it to other shells' behavior.  All other things being equal, consistency is good.  Is there a good reason for fish to be different in this important way?

`fish -c 'set i 1; for i in (seq 5); echo -n $i; end; echo $i'`
`123451`
`bash -c 'i=1; for i in $(seq 5); do echo -n $i; done; echo $i'`
`123455`
`dash -c 'i=1; for i in $(seq 5); do echo -n $i; done; echo $i'`
`123455`
`zsh -c 'i=1; for i in $(seq 5); do echo -n $i; done; echo $i'`
`123455`
`mksh -c 'i=1; for i in $(seq 5); do echo -n $i; done; echo $i'`
`123455`

**Comment by zanchey** [comment]
No. I think we'd take a fix.

**Comment by krader1961** [comment]
There is another reason to put the var in the enclosing scope. If you break out of the loop early it makes it possible to easily detect without having to introduce yet another var. However, this could only be changed in a major release and ideally we'd like to do more than one incompatible change if we're going to go to the trouble of making a major release.

**Comment by GReagle** [comment]
Following up on my earlier comment about being consistent with other shells, this is interesting.

This is [rc by Byron Rakitzis](http://tobold.org/article/rc) via [debian package "rc" 1.7.2-1](https://packages.debian.org/jessie/rc):
```
/usr/bin/rc -c 'i=1; for (i in `{seq 5}) {echo -n $i}; echo $i'
123455
```

This is [rc by Tom Duff](http://doc.cat-v.org/plan_9/4th_edition/papers/rc) from [Plan 9 from User Space](https://github.com/9fans/plan9port):
```
9 rc -c 'i=1; for (i in `{seq 5}) {echo -n $i}; echo $i'
123451
```

**Comment by krader1961** [comment]
You can fix this by cherry-picking commit cc3692fb from https://github.com/krader1961/fish.

**Comment by krader1961** [comment]
That's weird. I did not expect that fixing this in my fork would close this issue. I'll be more careful in the future about mentioning issues in this fork.
